diff --git a/axlearn/common/checkpointer_orbax.py b/axlearn/common/checkpointer_orbax.py
index 2eb205b..e44381e 100644
--- a/axlearn/common/checkpointer_orbax.py
+++ b/axlearn/common/checkpointer_orbax.py
@@ -321,11 +365,42 @@ class OrbaxCheckpointer(BaseCheckpointer):
 
         cfg: OrbaxCheckpointer.Config = self.config
 
+        replica_axis_index = 1
+
+        if cfg.enable_single_replica_ckpt_restoring:
+            array_handler = ocp.type_handlers.SingleReplicaArrayHandler(
+                replica_axis_index=replica_axis_index, primary_replica_id=0
+            )
+            ocp.type_handlers.register_type_handler(jax.Array, array_handler, override=True)
+
         def _restore_args(x: Any) -> ocp.RestoreArgs:
             if isinstance(x, (Tensor, TensorSpec)):
-                return ocp.checkpoint_utils.construct_restore_args(
+                arg = ocp.checkpoint_utils.construct_restore_args(
                     jax.ShapeDtypeStruct(shape=x.shape, dtype=x.dtype, sharding=x.sharding)
                 )
+                if cfg.enable_single_replica_ckpt_restoring and isinstance(
+                    arg, ocp.type_handlers.ArrayRestoreArgs
+                ):
+                    pspec = x.sharding.spec
+                    mesh = x.sharding.mesh
+                    devices = _replica_devices(mesh.devices, replica_axis_index)
+
+                    replica_mesh = jax.sharding.Mesh(devices, mesh.axis_names)
+                    single_replica_sharding = jax.sharding.NamedSharding(replica_mesh, pspec)
+
+                    arg = ocp.type_handlers.SingleReplicaArrayRestoreArgs(
+                        restore_type=arg.restore_type,
+                        dtype=arg.dtype,
+                        mesh=arg.mesh,
+                        mesh_axes=arg.mesh_axes,
+                        sharding=arg.sharding,
+                        global_shape=arg.global_shape,
+                        shape=arg.shape,
+                        strict=arg.strict,
+                        # the only added field
+                        single_replica_sharding=single_replica_sharding,
+                    )
+                return arg
             elif isinstance(x, tf.data.Iterator):
                 return _TfIteratorHandler.RestoreArgs(item=x)
             elif _GRAIN_INSTALLED and isinstance(x, _GrainIterator):
@@ -349,6 +424,17 @@ class OrbaxCheckpointer(BaseCheckpointer):
                 raise ValueError(f"Failed to restore at step {step}.") from e
             logging.info("Could not find any completed checkpoints under %s: %s", cfg.dir, e)
             return None, state  # Return the input state.
+        finally:
+            if cfg.enable_single_replica_ckpt_restoring:
+                ocp.type_handlers.register_type_handler(
+                    jax.Array,
+                    ArrayHandler(
+                        array_metadata_store=array_metadata_store_lib.Store(),
+                        use_replica_parallel=False,
+                        enable_write_sharding_file=False,
+                    ),
+                    override=True,
+                )
 
         restored_index = composite_state["index"]
         restored_state = composite_state["state"]
@@ -375,3 +461,29 @@ class OrbaxCheckpointer(BaseCheckpointer):
     def stop(self, *, has_exception: bool = False):
         """See `BaseCheckpointer.stop` for details."""
         self._manager.close()
+
+
+def _find_idx(array: np.ndarray, replica_axis_idx: int):
+    """Returns the index along given dimension that the current host belongs to."""
+    idx = None
+    for idx, val in np.ndenumerate(array):
+        if val.process_index == jax.process_index():
+            break
+    return idx[replica_axis_idx]
+
+
+def _replica_devices(device_array: np.ndarray, replica_axis_idx: int):
+    """Returns the devices from the replica that current host belongs to.
+
+    Replicas are assumed to be restricted to the first axis.
+
+    Args:
+      device_array: devices of the mesh that can be obtained by mesh.devices()
+      replica_axis_idx: axis dimension along which replica is taken
+
+    Returns:
+      devices inside the replica that current host is in
+    """
+    idx = _find_idx(device_array, replica_axis_idx)
+    replica_result = np.take(device_array, idx, axis=replica_axis_idx)
+    return np.expand_dims(replica_result, axis=replica_axis_idx)
